optimum/habana/transformers/generation/utils.py

```diff
@@ -1392,11 +1404,18 @@ def greedy_search(
                     break
 
             if bucket_size > 0:
-                # it will not have been padded if bucket_size > 0
-                params = next(inc)
-                input_ids, model_kwargs = self.update_model_kwargs_for_bucketing(
-                    params, input_ids, model_kwargs, pad_token_id, bucket_size, reduce_recompile
-                )
+                if not bucket_internal:
+                    # it will not have been padded if bucket_size > 0
+                    params = next(inc)
+                    input_ids, model_kwargs = self.update_model_kwargs_for_bucketing(
+                        params, input_ids, model_kwargs, pad_token_id, bucket_size, reduce_recompile
+                    )
+                else:
+                    # Calculate slice idx for kv cache. Breaking down the kv cache in the attention block helps to reduce computation time.
+                    idx = torch.div(model_kwargs.get("token_idx") - 1, bucket_size, rounding_mode="floor")
+                    if idx < (model_kwargs["kv_cache_len"] // bucket_size):
+                        cache_idx = (idx.item() + 1) * bucket_size
+                        model_kwargs["cache_idx"] = cache_idx
 
```

Comment:
@xt574chen  - this logic will work only when your total generated length is multiple of bucket size. For example consider an example of total length as 2060. So for tokens getting generated between 2048 and 2060, KV cache will be sliced till seq len 2048 and KV values between 2048 and 2060 won't be considered.

Please find updated logic below (spent a lot of time reviewing all the changes today):

if model_kwargs.get("token_idx") <= (model_kwargs["kv_cache_len"] // bucket_size) * bucket_size:
      idx = torch.div(model_kwargs.get("token_idx") - 1, bucket_size, rounding_mode="floor")
     cache_idx = (idx.item() + 1) * bucket_size
     model_kwargs["cache_idx"] = cache_idx
else:
     model_kwargs["cache_idx"] = model_kwargs["kv_cache_len"]

We can also avoid enhance a bit more for avoiding .item() call when the idx tensor is not changing. But lets avoid that for now maybe. We can push separate PR later. More importantly above logic needs to go in first so that the model logic works fine.

Reply:
@xt574chen  Further enhanced code can be as below. I will let you decide the best course of action.

```
#Declare prev_idx = None outside while loop.

if model_kwargs.get("token_idx") <= (model_kwargs["kv_cache_len"] // bucket_size) * bucket_size:
     idx = torch.div(model_kwargs.get("token_idx") - 1, bucket_size, rounding_mode="floor")
     if idx != prev_idx:
         cache_idx = (idx.item() + 1) * bucket_size
         model_kwargs["cache_idx"] = cache_idx
         prev_idx = idx
else:
     model_kwargs["cache_idx"] = model_kwargs["kv_cache_len"]
```

Reply:
I tested the recommended logic (without the enhancement) quite a bit and seems to be working fine. @xt574chen  - please test from your side and feel free to update anything.  But the original code has the issue as I highlighted in earlier comments.